Secondary eigenvalue capabilities using Krylov subspace methods

cjosey opened the pull request:
As per Will's request.

This modification allows the expression of the M and A operators of the  transport equation Ax = 1/k Mx as independent transport sweeps.  The method currently in OpenMOC represents a single operator on which a non-linear power iteration method is performed.

As separated operators, each can be cast as a SciPy LinearOperator and sparse matrix math can be performed.  In the example c5g7-krylov.py, I use a combination of GMRES and eigs (an IRAM implementation) to then solve for the first 6 eigenvalues and eigenvectors of the problem.

This method has a few caveats:
* The method is not yet derived for any boundary condition other than vacuum.
* It requires 2k+1 eigenvector-sized arrays to be stored, where k is the number of eigenvectors required.
* Eigs does not output convergence information.
* CMFD not only does not accelerate the method, but actively prevents it from working.
* It is slower than unaccelerated non-linear power iteration for a single eigenvalue
* GMRES must converge more tightly than IRAM, but I do not know by precisely how much.
* I implemented it assuming double precision.  It can probably be rewritten to read the FP_PRECISION type and adjust accordingly.
* Implementation was only done for CPUSolver

wbinventor commented:
Thanks @cjosey, this is a freaking cool addition to the code! I really love seeing how cleanly you turned the MOC kernel into a SciPy ``LinearOperator``. I think this is the best example yet to illustrate how powerful the OpenMOC Python interface can truly be.

I just pulled and ran this code for the new "c5g7-krylov.py" to compute the 50 largest eigenmodes. For other developers out there, I thought I'd throw a few of @cjosey's best paintings up here to illustrate what this code can do:

**First eigenvector, group 7 flux**
<img src="https://cloud.githubusercontent.com/assets/209492/8445966/b7b78c50-1f58-11e5-8cdd-be34597112b1.png" width="400px" />

**Tenth eigenvector, group 6 flux**
<img src="https://cloud.githubusercontent.com/assets/209492/8446025/3f54ad00-1f59-11e5-82b4-d80a6002e57a.png" width="400px" />

**Thirsty-sixth eigenvector, group 1 flux**
<img src="https://cloud.githubusercontent.com/assets/209492/8446036/6afb50e4-1f59-11e5-9036-a9c1216e7c1f.png" width="400px" />

**Fiftieth eigenvector, group 1 flux**
<img src="https://cloud.githubusercontent.com/assets/209492/8446044/8883e5c2-1f59-11e5-8c8d-e8a860553453.png" width="400px" />

I'll review this over the next few days, and perhaps submit my own PR to your fork with some modifications. In particular, I'd like to wrap up the SciPy code in "c5g7-krylov.py" (lines 383-466) into a general "KrylovSolver" class in a new ``openmoc.krylov`` Python module.

cjosey commented:
I've brought in @wbinventor's code.  It runs fine, but crashes after outputting the final result with a "*** Error in `python3': double free or corruption (!prev): 0x0000000028627780 ***" error.  The backtrace points to cpython and hdf5.  I'll run valgrind overnight to see if I can hope to isolate the issue.

wbinventor commented:
@cjosey make sure to sync this up with latest version of develop. It is
possible this issue was fixed by a couple of the recent PRs which handled
memory deallocation.
On Jul 21, 2015 2:17 PM, "Colin Josey" <notifications@github.com> wrote:

> I've brought in @wbinventor <https://github.com/wbinventor>'s code. It
> runs fine, but crashes after outputting the final result with a "*** Error
> in `python3': double free or corruption (!prev): 0x0000000028627780 ***"
> error. The backtrace points to cpython and hdf5. I'll run valgrind
> overnight to see if I can hope to isolate the issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/162#issuecomment-123480863>.
>

cjosey commented:
The version I am testing contains commit 42bf1d117e619d65e0bd73e873953bfd41ab4ec6, which is the latest commit on develop.  I redownloaded, reinstalled, and reran the simulation to the same results.

wbinventor commented:
I thought I should update the other developers on the changes made to @cjosey's original version of this code. The core methodology remains intact as it was, but it has been abstracted into a new ``openmoc.krylov.IRAMSolver`` submodule/class. In particular, this introduces the following changes:

* New ``openmoc.krylov.IRAMSolver`` submodule/class
* Made ``IRAMSolver`` work with both ``CPUSolver`` and ``GPUSolver`` (good speedup too)
* Made ``IRAMSolver`` work for both single or double precision
* New ``openmoc.plotter.plot_eigenmode_flux(...)`` routine
* Added timer report
* Removed "c5g7-unreflected.py"
* Simplified SWIG interface files (unrelated to Krylov)
* Simplified ``openmoc.___`` submodule imports (unrelated to Krylov)

wbinventor commented:
@PrezNattyGibbs or @geogunow can one of you take on this PR? Would be helpful to spread out the work load, I'll review @samuelshaner's CMFD and PRs. I fully reviewed this PR by reviewing and re-engineering @cjosey's original code, all of which has now been incorporated here.

cjosey commented:
I just added error messages if wrong boundary conditions are used.  I only did it in the 2D case, as I wasn't sure if in 2D, the 3D boundaries are ever initialized.

I read through the complete list of SciPy solvers.  Here is what the documentation says:
1. spsolve requires an explicit matrix.
2. bicg operates on any matrix, but requires Ax and A^Tx.
3. bicgstab states A must be hermitian positive definite matrix. [Note 1]
4. cg states A must be hermitian positive definite.
5. cgs states A must be real.
6. gmres operates on any matrix.
7. lgmres operates on any matrix, but can additionally take a starting guess [Note 2]
8. minres operates on symmetric matrices
9. qmr operates on real matrices, but requires Ax and A^Tx.

[Note 1] I currently suspect that this is a typo in the documentation.  The bicgstab method is designed for non-symmetric matrices.  Matlab's bicgstab will do it for non-symmetric matrices.
[Note 2] This starting guess may significantly improve performance.

Anyways, if we had transposes (I note mention of adjoint work going on elsewhere), we could enable bicg and qmr.  http://mathworld.wolfram.com/ConjugateGradientSquaredMethod.html mentions that using bicg would be better than cgs under the condition that A^Tx did not cost too much more to calculate, as cgs has problems with convergence.  Bicg would also help minimize memory usage, but gmres will basically always win in the raw flops requirement.

Secondly, because of the limitations above, I've removed cg as an option.  It only works on Hermitian matrices, which would only occur if upscatter from group A precisely equaled downscatter for group A-1 for all groups and (I think) the geometry is homogeneous.  Further, if this were the case, we could skip inverting the matrix altogether, as SciPy's eigs(A,M=M) will work if M is Hermitian due to some efficient mathematical trickery (Matlab, I think, inverts in the general eigenvalue problem).

Other than what I just fixed and the "more converged" issue (the inverse must be "more converged" than the tolerance we expect on eigenvalue, but I am yet sure by how much), I've yet to see problems.

wbinventor commented:
@cjosey the new adjoint mode has been introduced in #187. It *should* be pretty plug-n-play with the ``IRAMSolver``, albeit perhaps with a few modifications to ensure that ``IRAMSolver.computeEigenModes()`` transposes the production matrices. However, I think this small extension to compute adjoint eigenmodes can wait until both PRs are finalized, so that they can each be reviewed and merged in independently. But it would be interesting to know if the adjoint mode works for bicg and qmr.

cjosey commented:
I'd probably just set it up so that the LinearOperator has two functions, the `matvec` which we already use, and `rmatvec` for the transpose.  I'd then set it up that whenever we switch between evaluating `matvec` and `rmatvec`, we'd call `Material::transposeProductionMatrices()` or `Solver::resetMaterials()` as needed.

wbinventor commented:
Looks like the latest PRs may have created a merge conflict or two - @cjosey can you merge with develop when you get a chance?

cjosey commented:
I believe that it is fixed now.

geogunow commented:
@wbinventor do you think this is ready to merge?

wbinventor commented:
Yes

On Thu, Jul 30, 2015 at 11:43 AM, Geoffrey Gunow <notifications@github.com>
wrote:

> @wbinventor <https://github.com/wbinventor> do you think this is ready to
> merge?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/162#issuecomment-126430174>.
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com

wbinventor commented:
But someone besides @cjosey and I should test it out

On Thu, Jul 30, 2015 at 11:45 AM, Will Boyd <wbinventor@gmail.com> wrote:

> Yes
>
> On Thu, Jul 30, 2015 at 11:43 AM, Geoffrey Gunow <notifications@github.com
> > wrote:
>
>> @wbinventor <https://github.com/wbinventor> do you think this is ready
>> to merge?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/mit-crpg/OpenMOC/pull/162#issuecomment-126430174>.
>>
>
>
>
> --
> Will Boyd
> Nuclear Science & Engineering
> Massachusetts Institute of Technology
> Cell: 423.413.8469
> wbinventor@gmail.com
>



-- 
Will Boyd
Nuclear Science & Engineering
Massachusetts Institute of Technology
Cell: 423.413.8469
wbinventor@gmail.com

wbinventor commented:
@cjosey looks like this needs to be sync'ed with develop once again. Sorry for the hassle. Perhaps @geogunow or @PrezNattyGibbs can merge this PR in next since has been up here the longest.

cjosey commented:
I looked through the merge conflicts.  It's severe enough that some parts just need to be flat rewritten.  It'll be a while until I can get this working again.

wbinventor commented:
@cjosey look for my PR to your fork which resolves the merge conflicts.

cjosey commented:
While it does resolve the merge conflicts, there were a few code changes that never got ported to my components (OpenMP changes in calculation of scatter/fission source).  I'll clean those up in a minute or two, run some tests on the code, and then comment when I think it's "good".

cjosey commented:
I finally tested this on the GPU, and am getting good results.  Due to the communication between CPU Krylov and GPU operator, it is highly recommended that if one simulates on the GPU to run with as many angles as memory supports.  The reason being that CPU/communication time is so large that the GPU is doing nothing the majority of the time, and the number of unknowns (and thus quantity of communication) is not a function of angles.  It also only slightly changes the number of iterations.  Going from 4 angles to 32 only doubled runtime.

Performing a memory scan indicates a few issues elsewhere in the code (once I isolate them, I'll report them in a separate PR), but none specific to, or added by my code as far as I can tell.

I think it's good.

geogunow commented:
This PR will be merged after #187. Afterwards, this PR won't auto-merge so you'll need to update it once more after #187 is merged. Then I'll test the inputs, etc and get this merged in soon.

markreed commented:
Yes! When George Washington crossed the Delaware, the outcome was far from certain, but his zeal never faltered. We now stand before the gates of blessed nuclear harmony!




Sent from my mobile phone - please excuse lack of grammar

> On Aug 4, 2015, at 1:35 PM, Colin Josey <notifications@github.com> wrote:
> 
> I finally tested this on the GPU, and am getting good results. Due to the communication between CPU Krylov and GPU operator, it is highly recommended that if one simulates on the GPU to run with as many angles as memory supports. The reason being that CPU/communication time is so large that the GPU is doing nothing the majority of the time, and the number of unknowns (and thus quantity of communication) is not a function of angles. It also only slightly changes the number of iterations. Going from 4 angles to 32 only doubled runtime.
> 
> Performing a memory scan indicates a few issues elsewhere in the code (once I isolate them, I'll report them in a separate PR), but none specific to, or added by my code as far as I can tell.
> 
> I think it's good.
> 
> —
> Reply to this email directly or view it on GitHub.
>

geogunow commented:
I just ran all the sample input cases with this PR and they seem to check out. I'm going to do another quick review of the code before merging.

geogunow commented on openmoc/__init__.py:
```diff
@@ -14,12 +14,12 @@
 
-# Set a log file name using a date and time
+# Build a log file name using the date and time
+g = lambda x: x.zfill(2)
```
Can we get a more descriptive variable name than g?

geogunow commented on src/CPUSolver.cpp:
```diff
@@ -323,6 +398,91 @@ void CPUSolver::computeFSRSources() {
+      sigma_t = material->getSigmaT();
+
+      /* Compute scatter + fission source for group G */
+      for (int G=0; G < _num_groups; G++) {
```
Personally, I'm uneasy with the use of ``G`` as a variable name for the scatter source. I would much prefer something like ``g`` and ``g_prime`` or if we want brevity ``g`` and ``gp`` might be alright.

geogunow commented on src/CPUSolver.cpp:
```diff
@@ -323,6 +398,91 @@ void CPUSolver::computeFSRSources() {
+      sigma_t = material->getSigmaT();
+
+      /* Compute scatter + fission source for group G */
+      for (int G=0; G < _num_groups; G++) {
```
ditto

geogunow commented on src/Solver.h:
```diff
@@ -207,6 +254,16 @@ class Solver {
   /**
+   * @brief Computes the total fission source for each FSR and energy group.
+   */
+  virtual void computeFSRFissionSources() =0;
```
spacing for all these..... just search and replace ``=0;`` with ``= 0;``

geogunow commented on src/accel/cuda/GPUSolver.cu:
```diff
@@ -194,6 +194,105 @@ __global__ void computeFSRSourcesOnDevice(int* FSR_materials,
+      fission_source += scalar_flux(tid,e) * nu_sigma_f[e];
+
+    /* Set the reduced source for FSR r in each group G */
+    for (int G=0; G < *num_groups; G++) {
```
ditto on the use of ``G``

geogunow commented on src/accel/cuda/GPUSolver.cu:
```diff
@@ -194,6 +194,105 @@ __global__ void computeFSRSourcesOnDevice(int* FSR_materials,
+    sigma_t = curr_material->_sigma_t;
+
+    /* Compute total scattering source for this FSR in group G */
+    for (int G=0; G < *num_groups; G++) {
```
again, about the use of ``G``

geogunow commented:
Ok, I went through and had some comments regarding style. Once those are addressed this will be merged.

cjosey commented:
I have made the stylistic changes, and I have repeated the C5G7 tests.  It all looks clear as far as I can tell.

geogunow commented:
@cjosey I've seen it both ways, so just as long as you're consistent it should be fine. I see you just made another commit. Is this ready for a final review?

cjosey commented:
Yeah, it is.  I added that last one because I forgot that my source calculators were copy-pasted from the original routines, and the original routine used the G and g syntax.

wbinventor commented:
This looks good to me!
